lib/commands/media-projection.js

```diff
@@ -0,0 +1,271 @@
+import _ from 'lodash';
+import { waitForCondition } from 'asyncbox';
+import { util, fs, net, tempDir } from 'appium/support';
+import path from 'path';
+import B from 'bluebird';
+import { SETTINGS_HELPER_PKG_ID } from '../android-helpers';
+
+
+const commands = {};
+
+// https://github.com/appium/io.appium.settings#internal-audio--video-recording
+const DEFAULT_EXT = '.mp4';
+const RECORDING_STARTUP_TIMEOUT_MS = 3 * 1000;
+const RECORDING_STOP_TIMEOUT_MS = 3 * 1000;
+const MIN_API_LEVEL = 29;
+const RECORDING_SERVICE_NAME = `${SETTINGS_HELPER_PKG_ID}/.recorder.RecorderService`;
+const RECORDING_ACTIVITY_NAME = `${SETTINGS_HELPER_PKG_ID}/io.appium.settings.Settings`;
+const RECORDING_ACTION_START = `${SETTINGS_HELPER_PKG_ID}.recording.ACTION_START`;
+const RECORDING_ACTION_STOP = `${SETTINGS_HELPER_PKG_ID}.recording.ACTION_STOP`;
+const RECORDINGS_ROOT = `/storage/emulated/0/Android/data/${SETTINGS_HELPER_PKG_ID}/files`;
+
+
+async function uploadRecordedMedia (localFile, remotePath = null, uploadOptions = {}) {
+  if (_.isEmpty(remotePath)) {
+    return (await util.toInMemoryBase64(localFile)).toString();
+  }
+
+  const {user, pass, method, headers, fileFieldName, formFields} = uploadOptions;
+  const options = {
+    method: method || 'PUT',
+    headers,
+    fileFieldName,
+    formFields,
+  };
+  if (user && pass) {
+    options.auth = {user, pass};
+  }
+  await net.uploadFile(localFile, remotePath, options);
+  return '';
+}
+
+async function verifyMediaProjectionRecordingIsSupported (adb) {
+  const apiLevel = await adb.getApiLevel();
+  if (apiLevel < MIN_API_LEVEL) {
+    throw new Error(`Media projection-based recording is not available on API Level ${apiLevel}. ` +
+      `Minimum required API Level is ${MIN_API_LEVEL}.`);
+  }
+}
+
+
+class MediaProjectionRecorder {
+  constructor (adb) {
+    this.adb = adb;
+  }
+
+  async isRunning () {
+    const stdout = await this.adb.shell(['dumpsys', 'activity', 'services']);
```

Comment:
it seems like the result can be filtered by the package name/activity:
```
adb dumpsys activity services RECORDING_SERVICE_NAME
```

I haven't verified well, but potentially does this have slightly better performance rather than getting all services...?

Empty result:

```
ACTIVITY MANAGER SERVICES (dumpsys activity services)
  (nothing)
```

Reply:
changed